x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/classification/MulticlassConfusionMatrix.java

```diff
@@ -0,0 +1,174 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.core.ml.dataframe.evaluation.classification;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.search.aggregations.AggregationBuilder;
+import org.elasticsearch.search.aggregations.AggregationBuilders;
+import org.elasticsearch.search.aggregations.Aggregations;
+import org.elasticsearch.search.aggregations.bucket.composite.CompositeAggregation;
+import org.elasticsearch.search.aggregations.bucket.composite.TermsValuesSourceBuilder;
+import org.elasticsearch.xpack.core.ml.dataframe.evaluation.EvaluationMetricResult;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+
+/**
+ * {@link MulticlassConfusionMatrix} is a metric that answers the question:
+ *   "How many examples belonging to class X were classified as Y by the classifier?"
+ * for all the possible class pairs {X, Y}.
+ */
+public class MulticlassConfusionMatrix implements ClassificationMetric {
+
+    public static final ParseField NAME = new ParseField("multiclass_confusion_matrix");
+
+    private static final ObjectParser<MulticlassConfusionMatrix, Void> PARSER =
+        new ObjectParser<>(NAME.getPreferredName(), true, MulticlassConfusionMatrix::new);
+
+    public static MulticlassConfusionMatrix fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    private static final String AGG_NAME = NAME.getPreferredName();
+    private static final String ACTUAL_CLASS_FIELD = "actual_class";
+    private static final String PREDICTED_CLASS_FIELD = "predicted_class";
+    private static final int MAX_NUM_CLASSES = 100;
+
+    public MulticlassConfusionMatrix() {
+    }
+
+    public MulticlassConfusionMatrix(StreamInput in) throws IOException {
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME.getPreferredName();
+    }
+
+    @Override
+    public String getName() {
+        return NAME.getPreferredName();
+    }
+
+    @Override
+    public final List<AggregationBuilder> aggs(String actualField, String predictedField) {
+        return List.of(
+            AggregationBuilders.composite(
+                AGG_NAME,
+                List.of(
+                    new TermsValuesSourceBuilder(ACTUAL_CLASS_FIELD).field(actualField),
+                    new TermsValuesSourceBuilder(PREDICTED_CLASS_FIELD).field(predictedField)
+                )
+            ).size(MAX_NUM_CLASSES * MAX_NUM_CLASSES)
```

Comment:
This is an interesting hard limit to have. 

@tveasey when it comes to classification, do you think we should support > 100 classes?

@przemekwitek if we do need to support > 100 classes, I think chaining together callbacks to scroll through the composite aggregation would be necessary. It is not overly complicated, but may cause some frustrating refactoring in the search execution.

Reply:
There are certainly cases where people would have more than 100 classes, but I think they'll be rare. We could consider this as an enhancement